{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "%run lgs00_general_functions.ipynb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Some necessary variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "identities = {frozenset(['amac','av']):82, frozenset(['cgre','av']):41, frozenset(['pplu','av']):18,\n",
    "              frozenset(['amac','cgre']):43, frozenset(['amac','pplu']):17, frozenset(['cgre','pplu']):19,\n",
    "        frozenset(['amacV14L','av']):82, frozenset(['amacV14L','cgre']):43, frozenset(['amacV14L','pplu']):17,\n",
    "             frozenset(['amac','amacV14L']):100}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_figure_letters(figure, coordinates, **kwargs):\n",
    "    ax = figure.add_subplot(1,1,1)\n",
    "    ax = fig.add_subplot(1,1,1)\n",
    "    for side in ['top','bottom','right','left']:\n",
    "        ax.spines[side].set_color(None)\n",
    "    ax.tick_params(labelcolor='w', top=False, bottom=False, left=False, right=False)\n",
    "    \n",
    "    for letter in coordinates:\n",
    "        plt.text(coordinates[letter][0], coordinates[letter][1], letter, weight='bold', **kwargs)\n",
    "    return ax"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure 2: effects of mutations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Side by side slices of landscape peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cheesescapes(n=9, col='brightness', cmap='BuGn', pad=0.15, center=(0,0)):\n",
    "    cm = plt.get_cmap(cmap)\n",
    "    \n",
    "    plt.scatter([center[0]]*2,[center[1]]*2, c=[0,1], cmap=cmap)\n",
    "    plt.colorbar(orientation='vertical', label='Median brightness', pad=pad)\n",
    "    \n",
    "    medians = {}\n",
    "    for i in range(n):\n",
    "        n_mask = data_aa_scaled['n_mut']==i\n",
    "        medians[i] = [data_aa_scaled[amac_mask & n_mask][col].median() * 255, \n",
    "                      data_aa_scaled[cgre_mask & n_mask][col].median() * 255,\n",
    "                      data_aa_scaled[pplu_mask & n_mask][col].median() * 255,\n",
    "                      data_aa_scaled[av_mask & n_mask][col].median() * 255]\n",
    "    x = 2\n",
    "    for i in range(n)[::-1]:\n",
    "        if i==n-1:\n",
    "            labels = ['\\n\\n\\namacGFP','cgreGFP','ppluGFP2','avGFP']\n",
    "        else:\n",
    "            labels = ['','','','']\n",
    "        c = cm([int(j) for j in medians[i]])\n",
    "        plt.pie([1,1,1,1], radius=(x+i)/10, colors=c, labels=labels, wedgeprops={'linewidth':0.5,'edgecolor':'w'},\n",
    "               startangle=-45, center=center, textprops={'fontsize': 12})\n",
    "        \n",
    "        c2 = cm(int(255 - medians[i][0]))\n",
    "        if i == 8:\n",
    "            plt.text((x+i)/10 - 0.1 + center[0], 0 + center[1], '8 mutations',\n",
    "                horizontalalignment='left', verticalalignment='center', color=c2, weight='semibold')\n",
    "        else:\n",
    "            plt.text((x+i)/10 - 0.1 + center[0], 0 + center[1], i,\n",
    "                horizontalalignment='left', verticalalignment='center', color=c2, weight='semibold')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Observed fluorescent vs expected fluorescent"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ratio_predicted2observed_fluo(dataset, n_mut, threshold=0.4, **kwargs):\n",
    "    '''Returns the fraction of genotypes observed to be bright, from among those predicted to be bright\n",
    "    (measure of negative epistasis), as well as the fraction of genotypes observed to be dark,\n",
    "    from among those predicted to be dark (measure of positive epistasis).'''\n",
    "    \n",
    "    n_mut_mask = dataset['n_mut']==n_mut\n",
    "    bright_prediction_mask = dataset['expected_effect']>=threshold\n",
    "    dark_prediction_mask = dataset['expected_effect']<threshold\n",
    "    is_bright_mask = dataset['measured_effect']>=threshold\n",
    "    \n",
    "    bright_predicted = dataset[n_mut_mask & bright_prediction_mask]\n",
    "    bright_real = dataset[n_mut_mask & is_bright_mask & bright_prediction_mask]\n",
    "    dark_predicted = dataset[n_mut_mask & dark_prediction_mask]\n",
    "    dark_real = dataset[n_mut_mask & ~is_bright_mask & dark_prediction_mask]\n",
    "    \n",
    "    return len(bright_real)/len(bright_predicted)\n",
    "\n",
    "\n",
    "def plot_predicted_bright_vs_real(dataset, threshold=False):\n",
    "    stds = {'amac' : 0.03148308406723709, 'cgre' : 0.02852658185221242, 'pplu' : 0.027489382693326424,\n",
    "          'amacV14L' : 0.029670382945273337, 'av' : 0.0817635644945253}\n",
    "\n",
    "#     brightness_thresholds = {'amac':np.log10(3000), 'cgre':np.log10(6000), 'pplu':np.log10(5000), 'av':2, \n",
    "#                              'amacV14L':np.log10(3000), 'amacV14V':np.log10(3000)}\n",
    "    \n",
    "\n",
    "\n",
    "    n = 8\n",
    "    name = {'amac':'amacGFP', 'av':'avGFP', 'cgre':'cgreGFP', 'pplu':'ppluGFP2'}\n",
    "    \n",
    "    for gene in ['amac', 'cgre', 'pplu','av' ,]:\n",
    "        if threshold == False:\n",
    "            threshold = gate_borders_log[gene][0] - ref_wt_log[gene+'GFP']\n",
    "            \n",
    "        plt.plot([100 * get_ratio_predicted2observed_fluo(dataset[eval(gene+'_mask')], n_mut=i, \n",
    "                            threshold=threshold) for i in range(1,n)],\n",
    "                 color=colors[gene],label=name[gene], linewidth=3)\n",
    "            \n",
    "        plt.legend(loc='lower left', frameon=False)\n",
    "        plt.ylim(0,105)\n",
    "        plt.xticks(range(n-1), range(1,n))\n",
    "        label_plot_axis(x='Number of amino acid substitutions',\n",
    "                        y='Observed v. expected\\nfunctional genotypes (%)',)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Buried vs exposed single mutants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_buried_vs_exposed_violins(cm='mako' ):\n",
    "    c2 = ListedColormap(sns.color_palette(cm, 256))\n",
    "    colors = [c2(200), c2(100)]\n",
    "    \n",
    "    sns.violinplot(data=data_aa_scaled[~wt_mask & singles_mask], x='gene', y='brightness', \n",
    "                   saturation=1,\n",
    "                   palette=colors, width=1,\n",
    "                   order=['amacGFP','cgreGFP','ppluGFP','avGFP',], \n",
    "                   hue='has_buried_mutation', col='gene', split=True, linewidth=0, scale='area', cut=0)\n",
    "    \n",
    "    legendary(colors + ['w'], ['Exposed sites', 'Buried sites', 'Synonymous'], \n",
    "              colors + ['k'],\n",
    "              ncol=3, loc='upper center')\n",
    "    plt.ylim(-0.05, 1.4)\n",
    "    plt.yticks([0,0.25,0.5,0.75,1,1.25], [0,0.25,0.5,0.75,1,1.25])\n",
    "    \n",
    "    plot_half_violin([data_nt[wt_mask_nt & genemask]['scaled_brightness'] for genemask in [amac_mask_nt, cgre_mask_nt,\n",
    "                                                                                 pplu_mask_nt, av_mask_nt]],\n",
    "                    side='left', color='w', show_medians=False, alpha=1, widths=0.3)\n",
    "    plt.xticks([0,1,2,3], ['amacGFP', 'cgreGFP', 'ppluGFP2', 'avGFP'])\n",
    "    label_plot_axis(y='Fluorescence of\\nsingle mutants', )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Plot violins of whole libraries and controls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def plot_violin_libraries():\n",
    "#     sns.violinplot(data = data_aa_scaled[(data_aa_scaled.gene!='amacV14LGFP')\n",
    "#                         &(data_aa_scaled.aa_genotype_pseudo.str.contains('G69|Y68|R99'))],y='brightness',x='gene', \n",
    "#                         color='k', linewidth=0, alpha=0.5, cut=0, split=True)\n",
    "    \n",
    "#     sns.violinplot(data = data_aa_scaled[(data_aa_scaled.gene!='amacV14LGFP')], y='brightness', x='gene',\n",
    "#                         palette=[colors[gene] for gene in ['amac','cgre','pplu','av']], linewidth=0, cut=0, )\n",
    "    \n",
    "#     sns.violinplot(data = data_nt[(data_nt.gene!='amacV14LGFP')&(data_nt.aa_genotype_pseudo=='wt')], cut=0,\n",
    "#                y='scaled_brightness',x='gene', linewidth=1, color='w', hatch='//', inner=None, edgecolor='k')\n",
    "    \n",
    "#     patch_violinplot([None]*8+[colors[gene] for gene in ['amac','cgre','pplu','av']])\n",
    "    \n",
    "#     legendary(['k','w'], ['Chromophore mutants','Wild-type proteins', ], edges=['k','k'],\n",
    "#          loc = 'upper center', ncol=2, fontsize=10)\n",
    "    \n",
    "#     label_plot_axis(x='', y='Fluorescence of libraries')\n",
    "#     plt.ylim(-0.05, 1.4)\n",
    "#     plt.yticks([0,0.25,0.5,0.75,1,1.25],[0,0.25,0.5,0.75,1,1.25])\n",
    "#     plt.xticks([0,1,2,3],['amacGFP', 'cgreGFP', 'ppluGFP2', 'avGFP'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def patch_violinplot(facecolors, edgecolors, ):\n",
    "    from matplotlib.collections import PolyCollection\n",
    "    ax = plt.gca()\n",
    "    violins = [art for art in ax.get_children() if isinstance(art, PolyCollection)]\n",
    "#     colors = sns.color_palette(palette, n_colors=n) * (len(violins)//n)\n",
    "    for i in range(len(violins)):\n",
    "        violins[i].set_edgecolor(edgecolors[i])\n",
    "        violins[i].set_facecolor(facecolors[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_buried_vs_exposed_violins_singles(gene):\n",
    "    \n",
    "    ax = sns.violinplot(data=data_aa[~wt_mask & singles_mask & (data_aa.gene==gene)], \n",
    "                        x='gene', y='brightness', \n",
    "                   saturation=1,width=1,\n",
    "                   hue='has_buried_mutation', col='gene', split=True, \n",
    "               linewidth=2, inner=None, scale='area', cut=0)\n",
    "\n",
    "    patch_violinplot(edgecolors=[None, colors[gene]], facecolors=[colors[gene],'w'])\n",
    "    \n",
    "    legendary(['k', 'w'], ['Exposed sites', 'Buried sites'], edges=['k','k'],\n",
    "              ncol=2, loc='upper center')\n",
    "    \n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ax.spines['bottom'].set_visible(False)\n",
    "    ax.spines['left'].set_position('center')\n",
    "    ax.get_legend().remove()\n",
    "    \n",
    "    yticks = {'amacGFP' : [3,3.5,4,4.5], 'cgreGFP':[3,3.5,4,4.5], \n",
    "              'ppluGFP':[3,3.5,4,4.5], 'avGFP':[1.5,2,2.5,3,3.5,4]}\n",
    "    plt.yticks(yticks[gene],yticks[gene])\n",
    "    \n",
    "    label_plot_axis(y='Fluorescence of single mutants', )\n",
    "    plt.xticks([0],[names[gene]],fontsize=14)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_library_distributions(gene, nt_data, aa_data, ax, \n",
    "                               y1_label='Number of genotypes: Library', \n",
    "                               y2_label='Number of genotypes:\\nWTs, chromophore mutants'):\n",
    "    library = aa_data[(aa_data.gene==gene)]\n",
    "    wts = nt_data[(nt_data.gene==gene) & (nt_data.aa_genotype_native=='wt')]\n",
    "    \n",
    "    nonsense_mask = (aa_data.aa_genotype_pseudo.str.contains('\\*')) & ~(aa_data.aa_genotype_pseudo.str.contains('\\*246'))\n",
    "    chromomuts = aa_data[(aa_data.gene==gene) & ~nonsense_mask & (aa_data.n_mut<=20)\n",
    "                         & (aa_data.aa_genotype_pseudo.str.contains('G69|Y68|R99'))]\n",
    "    \n",
    "    sns.histplot(data = library['brightness'], \n",
    "                        bins=50, element='poly', color=colors[gene], fill=True, linewidth=2, )\n",
    "    \n",
    "    plt.ylim(0,4500)\n",
    "    plt.yticks([])\n",
    "    plt.ylabel('')\n",
    "    plt.xlabel('Fluorescence (log)', fontsize=12)\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "\n",
    "    ax.twinx()\n",
    "    wtmean = round(wts['log_brightness'].mean(),1)\n",
    "    wtstd = round(wts['log_brightness'].std(),3)\n",
    "    sns.histplot(data = wts['log_brightness'],\n",
    "                        bins=10, element='poly', color='w', fill=True, linewidth=2, edgecolor=colors[gene])\n",
    "    \n",
    "\n",
    "    sns.histplot(data = chromomuts['brightness'],\n",
    "                        bins=50, element='poly', color='k', fill=True, linewidth=0, )\n",
    "    plt.ylabel('')\n",
    "    if gene=='avGFP':\n",
    "        plt.ylim(0,200)\n",
    "    elif gene=='cgreGFP':\n",
    "        plt.ylim(0,900)\n",
    "        \n",
    "    \n",
    "        \n",
    "    y2_ticks = {'amacGFP':range(0,500,100), 'cgreGFP':range(0,900,200), \n",
    "               'ppluGFP':range(0,1600,400), 'avGFP':range(0,200,50)}\n",
    "    plt.yticks([])\n",
    "    legendary(colors = [colors[gene],'w','k'], edges = [colors[gene],colors[gene], 'k'],\n",
    "              labels = [f'Library\\n(n = {len(library)})', f'WT protein \\n(n = {len(wts)})',\n",
    "                       f'Mut. chromophore\\n(n = {len(chromomuts)})'], loc='upper left',\n",
    "             )\n",
    "#     plt.setp(legend.get_title(),fontsize=14)\n",
    "    plt.title(names[gene], fontsize=13)    \n",
    "    plt.sca(ax)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure 3: Neutral-to-deleterious change vs sequence distance"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Mutations which are neutral in one background and become deleterious in another"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_fraction_changed_effect(data, gene1, gene2, neutral_threshold, bad_threshold, mode='n2b'):\n",
    "    # n2b = neutral in gene1, becomes deleterious in gene2\n",
    "    # n2n =  neutral in gene1, remains neutral in gene2\n",
    "    # b2n = deleterious in gene1, becomes neutral in gene1\n",
    "    # b2b = deleterious in gene1, remains deleterious in gene2\n",
    "    assert mode in ['n2b', 'n2n', 'b2n', 'b2b']\n",
    "    \n",
    "    df = data[['effect_in_'+gene1, 'effect_in_'+gene2]].dropna() # muts in df are observed in both genes\n",
    "    df = df.reset_index()\n",
    "    \n",
    "    neutral_in_gene1_mask = df['effect_in_'+gene1] > neutral_threshold\n",
    "    neutral_in_gene2_mask = df['effect_in_'+gene2] > neutral_threshold\n",
    "    bad_in_gene1_mask = df['effect_in_'+gene1] < bad_threshold\n",
    "    bad_in_gene2_mask = df['effect_in_'+gene2] < bad_threshold\n",
    "    \n",
    "    if mode =='n2b':\n",
    "        return len(df[neutral_in_gene1_mask & bad_in_gene2_mask]) / len(df[neutral_in_gene1_mask])\n",
    "    elif mode == 'n2n':\n",
    "        return len(df[neutral_in_gene1_mask & neutral_in_gene2_mask]) / len(df[neutral_in_gene1_mask])\n",
    "    elif mode == 'b2n':\n",
    "        return len(df[bad_in_gene1_mask & neutral_in_gene2_mask]) / len(df[bad_in_gene1_mask])\n",
    "    elif mode == 'b2b':\n",
    "        return len(df[bad_in_gene1_mask & bad_in_gene2_mask]) / len(df[bad_in_gene1_mask])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_fraction_changed_effect(data, neutral_threshold, bad_threshold, mode, style, se=75, sf=10, **kwargs):\n",
    "    genes = ['amacV14L', 'amac', 'cgre', 'pplu', 'av', ]\n",
    "    shapes = {'amac':'P', 'amacV14L':'X', 'cgre':'o', 'pplu':'s', 'av':'d'}\n",
    "    gene_names = {'amac':'amacGFP', 'cgre':'cgreGFP', 'av':'avGFP', 'pplu':'ppluGFP2', 'amacV14L':'amacGFP:V14L'}\n",
    "    for gene1 in genes:\n",
    "        for gene2 in genes:\n",
    "            if gene1 != gene2:\n",
    "                x = 100 - identities[frozenset([gene1,gene2])]\n",
    "                y = get_fraction_changed_effect(data, gene1, gene2, \n",
    "                                                neutral_threshold, bad_threshold, mode)\n",
    "                if style == 'simple':\n",
    "                    plt.scatter(x, y, s=sf, color=colors[gene2], edgecolor='w', **kwargs)\n",
    "                elif style=='orbits':\n",
    "                    plt.scatter(x, y, s=se, facecolor=(0,0,0,0), marker='o',\n",
    "                                edgecolor=colors[gene1], linewidth=1.5, linestyle='--')\n",
    "                    plt.scatter(x, y, s=sf, color=colors[gene2], linewidth=0, marker='o')\n",
    "                    \n",
    "                elif style=='shapes':\n",
    "                    plt.scatter(x, y, s=sf, color=colors[gene2], edgecolor='w', linewidth=1, marker=shapes[gene1])\n",
    "                else:\n",
    "                    plt.scatter(x, y, s=sf, color=style, edgecolor='w', **kwargs)\n",
    "                   \n",
    "                \n",
    "    plt.xticks([0,18,58,82])\n",
    "#     label_plot_axis(x = 'Sequence distance', y = 'Fraction of mutations')\n",
    "    plt.xlim(-2, 85)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_overall_fractions_cross_gene_effects(gene1, gene2, neutral_threshold, bad_threshold):\n",
    "    df = df_effects_singles[['effect_in_'+gene1, 'effect_in_'+gene2]].dropna() # muts must be obsered in both genes\n",
    "    df = df.reset_index()\n",
    "    \n",
    "    wts = {'amac' : 3.974485547944876, 'cgre' : 4.49832043653411, 'pplu' : 4.226492294827216,\n",
    "          'amacV14L' : 3.9577171946556176, 'av' : 3.726358457799625}\n",
    "    \n",
    "    stds = {'amac' : 0.03148308406723709, 'cgre' : 0.02852658185221242, 'pplu' : 0.027489382693326424,\n",
    "          'amacV14L' : 0.029670382945273337, 'av' : 0.0817635644945253}\n",
    "    \n",
    "    neutral_in_both_mask = (df['effect_in_'+gene1] > -stds[gene1]*neutral_threshold) & (\n",
    "                                        df['effect_in_'+gene2] > -stds[gene2]*neutral_threshold)\n",
    "    \n",
    "    bad_in_both_mask = (df['effect_in_'+gene1] < -stds[gene1]*bad_threshold) & (\n",
    "                                        df['effect_in_'+gene2] < -stds[gene2]*bad_threshold)\n",
    "    \n",
    "    changed_effect_mask = ((df['effect_in_'+gene1] > -stds[gene1]*neutral_threshold) & \n",
    "                            (df['effect_in_'+gene2] < -stds[gene2]*bad_threshold)) | (\n",
    "                            (df['effect_in_'+gene2] > -stds[gene2]*neutral_threshold) & \n",
    "                            (df['effect_in_'+gene1] < -stds[gene1]*bad_threshold))\n",
    "    \n",
    "#     print('neutral in both', len(df[neutral_in_both_mask]) / len(df))\n",
    "#     print('bad in both', len(df[bad_in_both_mask]) / len(df))\n",
    "#     print('changed effect', len(df[changed_effect_mask]), len(df))\n",
    "\n",
    "    return {'nn' : len(df[neutral_in_both_mask]) / len(df) ,\n",
    "           'bb' : len(df[bad_in_both_mask]) / len(df), 'bn' : len(df[changed_effect_mask]) / len(df)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_bars_effectchange_epimaintain():\n",
    "    mako = ListedColormap(sns.color_palette('mako', 256))\n",
    "\n",
    "    shares = {}\n",
    "    fractions = {}\n",
    "    pairs = [('amac','amacV14L'), ('amac','av'),('amacV14L','av'),\n",
    "                 ('amac','cgre'), ('amacV14L', 'cgre'), ('av','cgre'),\n",
    "             ('amac','pplu'), ('amacV14L','pplu'), ('av','pplu'), ('cgre','pplu')]\n",
    "\n",
    "    for pair in pairs:\n",
    "        fractions[pair] = get_overall_fractions_cross_gene_effects(pair[0],pair[1],2,5)\n",
    "        shares[pair] = get_overall_epistasis_overlap(pair[0], pair[1], e=0.3)\n",
    "\n",
    "    x = np.array([0, 2,3, 5,6,7, 9,10,11,12])\n",
    "\n",
    "    plt.bar(x-0.2, [100*fractions[pair]['bn'] for pair in pairs], width=0.3, color=mako(75))\n",
    "    plt.ylabel('Neutrality change\\nof single mutations (%)', color=mako(75), fontsize=13)\n",
    "    plt.ylim(0,17)\n",
    "\n",
    "    plt.xticks(x, [f'{names[pair[0]]},\\n{names[pair[1]]}' for pair in pairs], rotation=55, ha='center' , \n",
    "               fontsize=8, va='center_baseline')\n",
    "\n",
    "    for xi,l,div in zip([0,2.5,6,10.5], [0.5,1,1.5,2], ['0.4%', '18%', '57-59%', '81-83% divergence']):\n",
    "#         plt.plot([xi-l, xi+l], [15,15], color='k')\n",
    "        plt.text(xi, 18, div, horizontalalignment='center')\n",
    "    for xi in [1,4,8]:\n",
    "        plt.axvline(xi, linewidth=1, linestyle = ':', color='k')\n",
    "\n",
    "    plt.twinx()\n",
    "    plt.bar(x+0.2, [100*shares[pair] for pair in pairs], width=0.3, color=mako(175))\n",
    "    plt.ylabel('Maintenance of\\npairwise epistasis (%)', color=mako(175), fontsize=13)\n",
    "    plt.ylim(0,10)\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_bars_effectchange_epimaintain_vertical(ax1, ax2):\n",
    "    mako = ListedColormap(sns.color_palette('bone', 256))\n",
    "\n",
    "    shares = {}\n",
    "    fractions = {}\n",
    "    pairs = [('amac','amacV14L'), ('amac','av'),('amacV14L','av'),\n",
    "                 ('amac','cgre'), ('amacV14L', 'cgre'), ('av','cgre'),\n",
    "             ('amac','pplu'), ('amacV14L','pplu'), ('av','pplu'), ('cgre','pplu')]\n",
    "\n",
    "    for pair in pairs:\n",
    "        fractions[pair] = get_overall_fractions_cross_gene_effects(pair[0],pair[1],2,5)\n",
    "        shares[pair] = get_overall_epistasis_overlap(pair[0], pair[1], e=0.3)\n",
    "\n",
    "    x = np.array([0, 2,3, 5,6,7, 9,10,11,12])\n",
    "    plt.sca(ax1)\n",
    "    plt.barh(x*1.2, [100*fractions[pair]['bn'] for pair in pairs], height=0.8, color=mako(125), )\n",
    "    plt.xlabel('Neutrality change\\n(% of mutations)', fontsize=12)\n",
    "    plt.gca().invert_yaxis()\n",
    "    plt.yticks(x*1.2, [f'{names[pair[0]]},\\n{names[pair[1]]}' for pair in pairs], rotation=0, ha='right' , \n",
    "               fontsize=9, va='center_baseline')\n",
    "#     plt.xlabel('Percent')\n",
    "\n",
    "    \n",
    "    for xi in np.array([1,4,8])*1.2:\n",
    "        plt.axhline(xi, linewidth=1, linestyle = ':', color='k')\n",
    "\n",
    "    plt.sca(ax2)\n",
    "    plt.barh(x*1.2, [100*shares[pair] for pair in pairs], height=0.8, color=mako(125))\n",
    "    plt.xlabel('Maintenance of epistasis\\n(% pairwise positions)', fontsize=12)\n",
    "    plt.yticks(x*1.2, ['']*len(x), rotation=0, ha='right' , fontsize=8, va='center_baseline')\n",
    "    for xi in np.array([1,4,8])*1.2:\n",
    "        plt.axhline(xi, linewidth=1, linestyle = ':', color='k')\n",
    "    for xi,div in zip(np.array([0,2.5,6,10.5])*1.2, \n",
    "                        ['0.4%\\ndivergence', '18%\\ndivergence', '57-59%\\ndivergence', '81-83%\\ndivergence']):\n",
    "        plt.text(8, xi, div, horizontalalignment='left', verticalalignment='center', fontsize=10)\n",
    "#     plt.xlabel('Percent')\n",
    "    plt.gca().invert_yaxis()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Pairs which are epistatic in one background and remain epistatic in another background"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def get_epistatic_pairs_overlap(gene1, gene2, e):\n",
    "#     pairs_gene1 = {x for x in epistatic_pairs_pos[gene1] if abs(max(epistatic_pairs_pos[gene1][x]))>e\n",
    "#                                       and x in epistatic_pairs_pos[gene2]}\n",
    "#     pairs_gene2 = {x for x in epistatic_pairs_pos[gene2] if abs(max(epistatic_pairs_pos[gene2][x]))>e\n",
    "#                                       and x in epistatic_pairs_pos[gene1]}\n",
    "#     return len([x for x in pairs_gene1 if x in pairs_gene2]) / len(pairs_gene1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def plot_shared_epistatic_pairs(e=0.3, style='shapes', se=75, sf=10, **kwargs):\n",
    "#     genes = ['amacV14L', 'amac', 'cgre', 'pplu', 'av', ]\n",
    "#     shapes = {'amac':'P', 'amacV14L':'X', 'cgre':'o', 'pplu':'s', 'av':'d'}\n",
    "#     gene_names = {'amac':'amacGFP', 'cgre':'cgreGFP', 'av':'avGFP', 'pplu':'ppluGFP2', 'amacV14L':'amacGFP:V14L'}\n",
    "#     for gene1 in genes:\n",
    "#         for gene2 in genes:\n",
    "#             if gene1 != gene2:\n",
    "#                 x = 100 - identities[frozenset([gene1,gene2])]\n",
    "#                 y = get_epistatic_pairs_overlap( gene1, gene2, e)\n",
    "#                 if style == 'simple':\n",
    "#                     plt.scatter(x, y, s=sf, color=colors[gene2], edgecolor='w', **kwargs)\n",
    "#                 elif style == 'orbits':\n",
    "#                     plt.scatter(x, y, s=se, facecolor=(0,0,0,0), marker='o',\n",
    "#                                 edgecolor=colors[gene1], linewidth=1.5, linestyle='--')\n",
    "#                     plt.scatter(x, y, s=sf, color=colors[gene2], linewidth=0, marker='o')\n",
    "#                 elif style == 'shapes':\n",
    "#                     plt.scatter(x, y, s=sf, color=colors[gene2], edgecolor='w', linewidth=1, marker=shapes[gene1])\n",
    "#                 else:\n",
    "#                     plt.scatter(x, y, s=sf, color=style, edgecolor='w',**kwargs)\n",
    "                \n",
    "#     plt.xticks([0,18,58,82])\n",
    "# #     label_plot_axis(x = 'Sequence distance between backgrounds', y = 'Fraction of pairs')\n",
    "#     plt.xlim(-2, 85)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_overall_epistasis_overlap(gene1, gene2, e):\n",
    "    # get mutation pairs which are measured to be epistatic in one gene\n",
    "    # and measured as non-epistatic in another gene\n",
    "    \n",
    "    epairs1 = {x for x in epistatic_pairs_pos[gene1] if\n",
    "                                abs(max(epistatic_pairs_pos[gene1][x]))>e\n",
    "                            and x in epistatic_pairs_pos[gene2]}\n",
    "    \n",
    "    epairs2 = {x for x in epistatic_pairs_pos[gene2] if\n",
    "                                abs(max(epistatic_pairs_pos[gene2][x]))>e\n",
    "                            and x in epistatic_pairs_pos[gene1]}\n",
    "    \n",
    "    eboth = [x for x in epairs1 if x in epairs2]\n",
    "    esingle = [x for x in epairs1 if x not in epairs2] + [x for x in epairs2 if x not in epairs1]\n",
    "    \n",
    "    return len(eboth) / len(esingle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure 4: Results of neural net predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_predictions_data():\n",
    "    predictions = pd.read_csv(os.path.join(data_folder, 'predictions',\n",
    "                                          'experimentally_tested_predictions.csv'))\n",
    "    fiji_ctrls = {'cgre_wt': 246, 'cgre_neg': 5, 'amac_wt': 65, 'amac_neg': 4, 'pplu_wt': 191, 'pplu_neg': 4}\n",
    "    predictions['fiji_log_value'] = np.log10(predictions['fiji_value'])\n",
    "    predictions['fiji_log_scaled'] = predictions[['fiji_log_value', 'gene']].apply(lambda x:\n",
    "                                        (x[0] - np.log10(fiji_ctrls[x[1]+'_neg'])) / \n",
    "                            (np.log10(fiji_ctrls[x[1]+'_wt']) -  np.log10(fiji_ctrls[x[1]+'_neg'])), axis=1)\n",
    "    predictions['fiji_log_scaled'] = predictions['fiji_log_scaled'].apply(lambda x: x if x>0 else 0)\n",
    "    return predictions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Violin plots of measured data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_all_vs_neutral_muts(gene, y_axis, threshold, df_effects, color):\n",
    "    df = data_aa_scaled[(data_aa['gene']==gene+'GFP')]\n",
    "    plot_half_violin([df[df['n_mut']==i][y_axis] for i in range(1,9)], side='left', \n",
    "                     color=color[1], alpha=1, widths=0.9, chonkylines=True)\n",
    "    bads = df_effects[df_effects['effect_in_'+gene] < -threshold].copy()\n",
    "    bads['full_mutation'] = bads[['wt_state_'+gene, 'position', 'mutation']].apply(lambda x:\n",
    "                                            x[0] + str(x[1]) + x[2], axis=1)\n",
    "    bads = set(bads['full_mutation'])\n",
    "    bads = {x for x in bads if '*' not in x}\n",
    "\n",
    "    df = df[(df['gene']==gene+'GFP') & ~(df['aa_genotype_pseudo'].str.contains('|'.join(bads)))]\n",
    "\n",
    "    plot_half_violin([df[df['n_mut']==i][y_axis] for i in range(1,9)], side='right', alpha=1, \n",
    "                     color=color[0], widths=0.9, chonkylines=True)\n",
    "\n",
    "    label_plot_axis(x = 'Number of mutations', t= gene+'GFP fitness')\n",
    "\n",
    "    if gene!='av':\n",
    "        plt.axhline(gate_borders_scaled[gene][0], color='crimson', linewidth=1, linestyle='--',\n",
    "                   path_effects=[pe.Stroke(linewidth=3, foreground='w'), pe.Normal()])\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_predictions(gene, color,mode='full',**kwargs):\n",
    "    c = ListedColormap(sns.color_palette('mako', 256))\n",
    "    pos = [1,2,3,4,5,6,7,8,12,18,24,30,36,42,48]\n",
    "    plot_all_vs_neutral_muts(gene, 'brightness', 0.05, \n",
    "                                         df_effects_singles_scaled, color=color)\n",
    "\n",
    "    dummy = predictions[predictions['gene']==gene].copy()\n",
    "    dists = list(range(1,49))\n",
    "    for d in [6,12,18,24,30,36,42,48]:\n",
    "        dists.remove(d)\n",
    "    for d in dists:\n",
    "        dummy.loc[10000*d,'distance'] = d\n",
    "        dummy.loc[10000*d,'gene'] = gene\n",
    "    sns.swarmplot(data=dummy[dummy['gene']==gene], x='distance', y='fiji_log_scaled', color='k',**kwargs)\n",
    "\n",
    "    dists = [dummy[dummy['distance']==x]['fiji_log_scaled'].median() for x in [6,12,18,24,30,36,42,48]]\n",
    "    plt.plot([x-1 for x in [6,12,18,24,30,36,42,48]], dists, color='k', linestyle='--')\n",
    "    \n",
    "    plt.xticks([i-1 for i in pos], pos);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Suppl. Fig. 2: Distribution of libraries during sorting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\lgonzale\\Anaconda3\\lib\\site-packages\\FlowCytometryTools\\core\\docstring.py:56: MatplotlibDeprecationWarning: \n",
      "The dedent function was deprecated in Matplotlib 3.1 and will be removed in 3.3. Use inspect.cleandoc instead.\n",
      "  doc = dedent(doc)\n"
     ]
    }
   ],
   "source": [
    "from FlowCytometryTools import FCMeasurement\n",
    "from FlowCytometryTools import ThresholdGate, PolyGate, IntervalGate, QuadGate\n",
    "\n",
    "def transform_by_log10(dataset):\n",
    "    dataset = np.where(dataset<=1, 0, np.log10(dataset))\n",
    "    return dataset\n",
    "\n",
    "def determine_filename(gene, machine, ctrl):\n",
    "    path = os.path.join(data_folder, 'cell_sorting', 'fcs_files')\n",
    "    machine = 'GFP_' + machine + '__' if gene!='pplu' else 'GFP2_' + machine + '__'\n",
    "    if ctrl in ['ctrl', 'negctrl']:\n",
    "        name = 'negative_control'\n",
    "    else:\n",
    "        name = 'library'\n",
    "    filename = os.path.join(path, gene + machine + name + '.fcs')\n",
    "    return filename\n",
    "\n",
    "def load_fcs_data(filename):\n",
    "    fcs = FCMeasurement(ID=filename, datafile=filename)\n",
    "    cols = [x for x in fcs.data.columns if x!='Time']\n",
    "    fcs = fcs.transform(transform_by_log10, auto_range=False, use_spln=False, channels=cols)\n",
    "    for channel_name in ['GFP-A', 'FITC-A']:\n",
    "        if channel_name in fcs.data.columns:\n",
    "            fcs.data.rename(columns={channel_name:'GFP'}, inplace=True)\n",
    "    for channel_name in ['mCherry-A', 'PE-Texas Red-A']:\n",
    "        if channel_name in fcs.data.columns:\n",
    "            fcs.data.rename(columns={channel_name:'mKate2'}, inplace=True)\n",
    "    return fcs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_bacteria_gate = ThresholdGate(threshold=1, channel=['FSC-A'], region='above', name='is_bacteria_gate')\n",
    "\n",
    "# approximately, based on sorting data pdfs from VBCF\n",
    "mKate2_gate_A = IntervalGate(vert=(3.2, 3.4), channel=['mKate2'], region='in', name='mKate2_gate_A')\n",
    "mKate2_gate_B = IntervalGate(vert=(3.4, 3.6), channel=['mKate2'], region='in', name='mKate2_gate_B')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 271,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sortingplot_library_scatter(gene, machine, fontscale=1, savepath=False):\n",
    "    redgate = {'A' : {'amac':(3.18, 3.35), 'cgre':(3.14, 3.32), 'pplu':(3.18, 3.44)},\n",
    "              'B' : {'amac':(3.4, 3.6), 'cgre':(3.29, 3.57), 'pplu':(3.28, 3.53)}}\n",
    "    if gene=='pplu':\n",
    "        gate_borders = pd.read_csv(os.path.join(data_folder, 'cell_sorting', 'ppluGFP2__gate_border_values.txt'),\n",
    "                              sep='\\t', index_col=0)\n",
    "    else:\n",
    "        gate_borders = pd.read_csv(os.path.join(data_folder, 'cell_sorting', gene+'GFP__gate_border_values.txt'),\n",
    "                                  sep='\\t', index_col=0)\n",
    "    gate_borders = {'A': np.log10(np.array(gate_borders.head(7)['upper_limit'])),\n",
    "                  'B': np.log10(np.array(gate_borders.iloc[8:15]['upper_limit']))}\n",
    "    \n",
    "    library = load_fcs_data(determine_filename(gene,machine,''))\n",
    "    library = library.gate(is_bacteria_gate).data.copy().replace(0,np.nan).sample(10000)\n",
    "    library[''] = 'Whole library'\n",
    "    ctrl = load_fcs_data(determine_filename(gene,machine,'ctrl'))\n",
    "    ctrl = ctrl.gate(is_bacteria_gate).data.copy().replace(0,np.nan).sample(1500)\n",
    "    ctrl[''] = 'Non-fluorescent controls'\n",
    "    data=pd.concat([library,ctrl])\n",
    "    data['sorted'] = data['mKate2'].apply(lambda x: True if \n",
    "                                          redgate[machine][gene][0]<x<redgate[machine][gene][1] else False)\n",
    "    \n",
    "    ax = sns.jointplot(data=data, x='GFP', y='mKate2', hue='', \n",
    "                       alpha=0.4, s=5, edgecolor=None, palette=['darkgrey', 'k',], )\n",
    "    ax.set_axis_labels('GFP fluorescence (log)','mKate2 fluorescence (log)', fontsize=10+fontscale)\n",
    "    ax.ax_joint.legend(frameon=False)\n",
    "\n",
    "    plt.sca(ax.ax_joint)\n",
    "    plt.scatter(data[data.sorted==True]['GFP'], data[data.sorted==True]['mKate2'], color='crimson', s=4, \n",
    "                label='Sorted cells', alpha=0.4)\n",
    "    legendary(['lightgrey', 'grey', 'crimson'], \n",
    "              ['Whole library', 'Non-fluorescent controls', 'Sorted cells'], fontsize=10+fontscale,\n",
    "             title='%sGFP library: sorter %s' % (gene,machine), title_fontsize=12+fontscale, loc='upper left')\n",
    "    plt.ylim(0,5.5)\n",
    "    plt.xlim(0,5.5)\n",
    "    plt.yticks(fontsize=10+fontscale)\n",
    "    plt.xticks(fontsize=10+fontscale)\n",
    "    \n",
    "    plt.sca(ax.ax_marg_x)\n",
    "    sns.kdeplot(x=data[data.sorted==True]['GFP'], color='crimson')\n",
    "    for border in gate_borders[machine]:\n",
    "        plt.axvline(border, color='crimson', linestyle='--', linewidth=1)\n",
    "    \n",
    "#     plt.fill_between([0,6], redgate[0], redgate[1], facecolor='r', edgecolor='w', alpha=0.4, linewidth=0,\n",
    "#                      hatch='////', )\n",
    "    return ax\n",
    "    if savepath:\n",
    "        savefig(savepath, dpi=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Suppl. Fig. 4: Mutation effects by position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_positional_effects_heatmap(df=df_effects_singles, func=np.nanmedian, cm='Blues_r'):\n",
    "    effects_amac = get_effects_by_position('amac', df, positions='pseudo', func=func)\n",
    "    effects_cgre = get_effects_by_position('cgre', df, positions='pseudo', func=func)\n",
    "    effects_pplu = get_effects_by_position('pplu', df, positions='pseudo', func=func)\n",
    "    effects_av = get_effects_by_position('av', df, positions='pseudo', func=func)\n",
    "\n",
    "    font = 'Arial'\n",
    "    \n",
    "    to_plot = np.array([effects_av, effects_amac, effects_cgre, effects_pplu, ])\n",
    "        \n",
    "    hm = sns.heatmap(to_plot, \n",
    "               cmap=cm, yticklabels=['avGFP','amacGFP', 'cgreGFP', 'ppluGFP2', ], xticklabels=10,\n",
    "               cbar_kws={'label': 'Median mutation effect', 'pad':0.01, 'fraction':0.05})\n",
    "\n",
    "    y = 0.5\n",
    "    for gene in ['av','amac','cgre','pplu']:\n",
    "        ss = pseudify(eval(gene+'_ss_pymol'), gene)\n",
    "        plot_secondary_structure(ss, y=y, hel=0.05, arrow_width=0.1, linewidth=1, c='k')\n",
    "        y += 1\n",
    "    plt.scatter([x+0.5 for x in buried_pos], [0 for x in buried_pos], color='k', s=50, marker=2)\n",
    "\n",
    "    plt.pcolor(np.arange(len(effects_amac)+1), np.arange(len(to_plot)+1), \n",
    "               np.ma.masked_less(np.array([[np.nan]*67+[10,10,10]+[np.nan]*178 for i in range(4)]), 5), \n",
    "               hatch='///', alpha=0)\n",
    "\n",
    "    label_plot_axis(x='Aligned amino acid position', t='Median effects of single mutations, by position')\n",
    "    plt.xticks(fontname=font)\n",
    "    plt.yticks(rotation=0, fontname=font)\n",
    "    plt.ylim(len(to_plot)+0.1, -0.1)\n",
    "    return to_plot, hm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.image as mpimg\n",
    "def show_image(image_path):\n",
    "    plt.imshow(mpimg.imread(image_path))\n",
    "    plt.axis('off')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Suppl. Fig. 5: General epistasis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def doublemut_scatterplot(gene):\n",
    "    df = doublemuts[doublemuts['gene']==gene+'GFP'].copy()\n",
    "    df['abs_epistasis'] = abs(df['epistasis'])\n",
    "    df.sort_values('abs_epistasis', inplace=True)\n",
    "    x,y,z = 'mut1_effect','mut2_effect','epistasis'\n",
    "    plt.scatter(df[x], df[y], alpha=0.8,\n",
    "               c=df[z], cmap='Spectral', s=5,vmin=-2, vmax=2)\n",
    "    label_plot_axis(x='Effect of first mutation', y='Effect of second mutation', t=gene+'GFP')\n",
    "#     cbar = plt.colorbar()\n",
    "#     cbar.ax.set_ylabel('Epistasis')\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def positional_heatmap(gene, include_singlemut_e=True):\n",
    "    # assumes positions are pseudopisitions, not native\n",
    "    \n",
    "    length = len(eval(gene+'_wt'))\n",
    "    \n",
    "    if include_singlemut_e==False:\n",
    "        positional_max_e = np.zeros([length, length])\n",
    "        for i in range(247):\n",
    "            for j in range(247):\n",
    "                pos1 = pseudopos_to_nativepos[i][genekey[gene]]\n",
    "                pos2 = pseudopos_to_nativepos[j][genekey[gene]]\n",
    "                if str(i)+':'+str(j) in epistatic_pairs_pos[gene]:\n",
    "                    positional_max_e[pos1,pos2] = max(map(abs, epistatic_pairs_pos[gene][str(i)+':'+str(j)]))\n",
    "                    positional_max_e[pos2,pos1] = 20\n",
    "                    \n",
    "        sns.heatmap(positional_max_e, vmin=0, vmax=1, cmap='GnBu', square=True, mask= positional_max_e==0,\n",
    "                    cbar_kws={'label':'epistasis', 'shrink':1})\n",
    "        plt.ylim(0,length)\n",
    "        plt.xlim(0,length)\n",
    "    #     ax.invert_yaxis()\n",
    "        plt.xticks(np.array(range(11,length,25))+0.5, range(0,length,25) )\n",
    "        plt.yticks(np.array(range(11,length,25))+0.5, range(0,length,25), rotation=0)\n",
    "        \n",
    "    else:\n",
    "        positional_max_e = np.zeros([length+10, length+10])\n",
    "        position_effects = get_effects_by_position(gene, df_effects_singles, positions='native')\n",
    "        for i in range(247):\n",
    "            for j in range(247):\n",
    "                pos1 = pseudopos_to_nativepos[i][genekey[gene]]\n",
    "                pos2 = pseudopos_to_nativepos[j][genekey[gene]]\n",
    "                if str(i)+':'+str(j) in epistatic_pairs_pos[gene]:\n",
    "                    positional_max_e[pos1+10, 0:10] = position_effects[pos1]-10\n",
    "                    positional_max_e[0:10, pos1+10] = position_effects[pos1]-10\n",
    "                    positional_max_e[pos1+10,pos2+10] = max(map(abs, epistatic_pairs_pos[gene][str(i)+':'+str(j)]))\n",
    "                    positional_max_e[pos2+10,pos1+10] = 20\n",
    "                    \n",
    "        sns.heatmap(positional_max_e, mask= positional_max_e<=0, vmin=0, vmax=2, cmap='Blues',\n",
    "               square=True, cbar_kws={'label': 'epistasis', 'shrink':1},)\n",
    "        sns.heatmap(positional_max_e, mask= positional_max_e>=0, vmin=np.nanmin(position_effects)-10, \n",
    "                vmax=np.nanmax(position_effects)-10, cmap='Blues_r', cbar=False, square=True, )\n",
    "            #cbar_kws={'label': 'singlemut effect', 'shrink':0.75, 'orientation':'horizontal'})\n",
    "            \n",
    "        plt.ylim(0,length+10)\n",
    "        plt.xlim(0,length+10)\n",
    "    #     ax.invert_yaxis()\n",
    "        plt.xticks(np.array(range(11,length+10,25))+0.5, range(0,length,25) )\n",
    "        plt.yticks(np.array(range(11,length+10,25))+0.5, range(0,length,25), rotation=0)\n",
    "\n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def barplot_epistasis(gene, e):\n",
    "    neg = [len(data_aa[(data_aa['gene']==gene+'GFP') & (data_aa['epistasis'] < -e) & (data_aa['n_mut']==i)]) / \n",
    "         len(data_aa[(data_aa['gene']==gene+'GFP') & (data_aa['n_mut']==i) & (data_aa['epistasis'].notnull())])\n",
    "        for i in range(2,9)]\n",
    "    pos = [len(data_aa[(data_aa['gene']==gene+'GFP') & (data_aa['epistasis'] > e) & (data_aa['n_mut']==i)]) / \n",
    "         len(data_aa[(data_aa['gene']==gene+'GFP') & (data_aa['n_mut']==i) & (data_aa['epistasis'].notnull())])\n",
    "        for i in range(2,9)]\n",
    "\n",
    "    plt.bar(range(len(neg)), neg, color=colors[gene])\n",
    "    plt.bar(range(len(neg)), pos, bottom=neg, color=colors[gene], alpha=1, hatch='///', \n",
    "            edgecolor='w', linewidth=0)\n",
    "    plt.xticks(range(len(neg)), range(2,11))\n",
    "    plt.ylim(0,0.25)\n",
    "    label_plot_axis(t=gene+'GFP', x='Number of mutations', y='Fraction of genotypes\\nwith epistasis over %s' % e);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Suppl. Fig. 6: Distances between amino acid pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def distances_heatmap(gene, epistasis=0.3):\n",
    "    length = len(eval(gene+'_wt'))\n",
    "    name = {'amac':'amacGFP', 'av':'avGFP', 'cgre':'cgreGFP', 'pplu':'ppluGFP2'}\n",
    "    df = pd.read_csv(os.path.join(structure_folder, 'residue_distance_matrices', \n",
    "                                  name[gene]+'__minimal_distances_between_aa.csv'), index_col=0)\n",
    "    positional_max_e = np.zeros([length, length])\n",
    "    sites_to_ignore = [x for x in range(length) if x not in df.index]\n",
    "    \n",
    "    for i in range(length):\n",
    "        for j in range(length):\n",
    "            if str(i)+':'+str(j) in epistatic_pairs_pos[gene]:\n",
    "                positional_max_e[i,j] = max(map(abs, epistatic_pairs_pos[gene][str(i)+':'+str(j)]))\n",
    "\n",
    "    positional_max_e = np.delete(positional_max_e, sites_to_ignore, axis=1)\n",
    "    positional_max_e = np.delete(positional_max_e, sites_to_ignore, axis=0)\n",
    "\n",
    "    d = np.array(df)\n",
    "    sns.heatmap(data=d, square=True, cmap='mako', mask=np.tril(d), cbar=False,\n",
    "                cbar_kws={'label': 'Minimum distance between residues (A)', 'shrink':1, 'pad':0.15})\n",
    "    ax = sns.heatmap(data=positional_max_e, mask=positional_max_e<0.3,square=True, alpha=1, \n",
    "                cmap='hot', cbar=False, vmin=0, vmax=0.3)\n",
    "    plt.ylim(0,len(d))\n",
    "    plt.xlim(0,len(d))\n",
    "    plt.xticks(range(0, len(d), 20), df.index[0::20], fontsize=8)\n",
    "    plt.yticks(range(0, len(d), 20), df.index[0::20], rotation=0, fontsize=8)\n",
    "    ax.yaxis.set_label_position('right')\n",
    "    ax.yaxis.tick_right()\n",
    "        \n",
    "    label_plot_axis(x='Amino acid position',y='Amino acid position', fontsize_x=10, fontsize_y=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def setup_inset(ax):\n",
    "    c = ListedColormap(sns.color_palette('mako', 256))\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['bottom'].set_visible(False)\n",
    "    ax.spines['left'].set_position('center')\n",
    "    plt.xticks(fontsize=12)\n",
    "    label_plot_axis(y='Distance between residues (A)', fontsize_y=8)\n",
    "    ax.text(0.1, 50, 'No epistasis', color=c(175))\n",
    "    ax.text(0.1, 45, '|Epistasis| > 0.3', color=c(75))\n",
    "    plt.ylim(0,50)\n",
    "\n",
    "def plot_all_distance_maps(df = doublemuts, e = 0.3):\n",
    "    df['e_above_threshold'] = df['epistasis'].apply(lambda x: True if abs(x)>e else False)\n",
    "    fig = plt.figure(figsize=[10,10], dpi=200)\n",
    "\n",
    "    c = ListedColormap(sns.color_palette('mako', 256))\n",
    "\n",
    "    plt.subplot(2,2,1)\n",
    "    distances_heatmap('amac')\n",
    "    left, bottom, width, height = [0.02, 0.75, 0.14, 0.18]\n",
    "    ax1 = fig.add_axes([left, bottom, width, height])\n",
    "    ax = sns.violinplot(data=df[(df.gene=='amacGFP')], x='gene', y='distance', linewidth=0, cut=0,\n",
    "                  hue='e_above_threshold', split=True, palette=[c(175), c(75)], )\n",
    "    ax.get_legend().remove()\n",
    "    setup_inset(ax1)\n",
    "\n",
    "    plt.subplot(2,2,2)\n",
    "    distances_heatmap('cgre')\n",
    "    left, bottom, width, height = [0.52, 0.75, 0.14, 0.18]\n",
    "    ax2 = fig.add_axes([left, bottom, width, height])\n",
    "    ax = sns.violinplot(data=df[(df.gene=='cgreGFP')], x='gene', y='distance', linewidth=0, cut=0,\n",
    "                  hue='e_above_threshold', split=True, palette=[c(175), c(75)], )\n",
    "    ax.get_legend().remove()\n",
    "    setup_inset(ax2)\n",
    "\n",
    "    plt.subplot(2,2,3)\n",
    "    distances_heatmap('pplu')\n",
    "    left, bottom, width, height = [0.02, 0.25, 0.14, 0.18]\n",
    "    ax3 = fig.add_axes([left, bottom, width, height])\n",
    "    ax = sns.violinplot(data=df[(df.gene=='ppluGFP')], x='gene', y='distance', linewidth=0, cut=0,\n",
    "                  hue='e_above_threshold', split=True, palette=[c(175), c(75)], )\n",
    "    ax.get_legend().remove()\n",
    "    setup_inset(ax3)\n",
    "\n",
    "    plt.subplot(2,2,4)\n",
    "    distances_heatmap('av')\n",
    "    left, bottom, width, height = [0.52, 0.25, 0.14, 0.18]\n",
    "    ax4 = fig.add_axes([left, bottom, width, height])\n",
    "    ax = sns.violinplot(data=df[(df.gene=='avGFP')], x='gene', y='distance', linewidth=0, cut=0,\n",
    "                  hue='e_above_threshold', split=True, palette=[c(175), c(75)], )\n",
    "    ax.get_legend().remove()\n",
    "    setup_inset(ax4)\n",
    "\n",
    "    plt.tight_layout(pad=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def distances_violins(df, epistasis, palette='mako'):\n",
    "    c = ListedColormap(sns.color_palette(palette, 256))\n",
    "    df['e_above_threshold'] = df['epistasis'].apply(lambda x: True if abs(x)>epistasis else False)\n",
    "\n",
    "    sns.violinplot(data=df[(df.gene!='amacV14LGFP') & \n",
    "                                   (abs(df['epistasis'])>0)], \n",
    "                   x='gene', y='distance', hue='e_above_threshold', split=True, linewidth=0,\n",
    "                  palette=[c(75), c(175)], saturation=100, cut=0)\n",
    "    legendary([c(75), c(175)], ['No epistasis', 'Epistasis > %s' % epistasis], ncol=2, \n",
    "              loc='upper center')\n",
    "    label_plot_axis(y='Distance (A) between\\namino acid pairs', )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Suppl. Fig. 7: Protein stability"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def import_plate_reader_data():\n",
    "#     spectra_fluo = pd.read_csv('urea_spectra_fluo_7_longform.txt', sep='\\t', index_col=0)\n",
    "    spectra_fluo = pd.read_csv(os.path.join(structure_folder, 'stability_measurements',\n",
    "                                           'urea__fluorescence_spectra.csv'), index_col=0)\n",
    "#     spectra_abs = pd.read_csv('urea_spectra_abs_longform.txt', sep='\\t', index_col=0)\n",
    "    spectra_abs = pd.read_csv(os.path.join(structure_folder, 'stability_measurements',\n",
    "                                           'urea__absorbance_spectra.csv'), index_col=0)\n",
    "\n",
    "    \n",
    "    spectra_fluo['\\nTreatment'] = spectra_fluo['treatment'].apply(lambda x: '9M urea' if x=='urea' else 'control')\n",
    "    spectra_fluo['hours'] = spectra_fluo['t'] * 25 / 60 #each fluorescence spectrum measurement takes ~25min\n",
    "    spectra_fluo['Gene'] = spectra_fluo['gene'].apply(lambda x: 'amacGFP' if x=='amac' else 'cgreGFP' if x=='cgre'\n",
    "                                                 else 'ppluGFP' if x=='pplu' else 'avGFP' if x=='av'\n",
    "                                                 else 'amacGFP_V14L' if x=='amacV14L' else 'blank')\n",
    "    \n",
    "    spectra_abs['\\nTreatment'] = spectra_abs['treatment'].apply(lambda x: '9M urea' if x=='urea' else 'control')\n",
    "    spectra_abs['hours'] = spectra_abs['t'] * 40 / 60 #each absorbance spectrum measurement takes ~40min\n",
    "    spectra_abs['Gene'] = spectra_abs['gene'].apply(lambda x: 'amacGFP' if x=='amac' else 'cgreGFP' if x=='cgre'\n",
    "                                                 else 'ppluGFP' if x=='pplu' else 'avGFP' if x=='av'\n",
    "                                                 else 'amacGFP_V14L' if x=='amacV14L' else 'blank')\n",
    "    return spectra_abs, spectra_fluo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_spectra(gene, treatment, normed=True):\n",
    "    if normed==True:\n",
    "        y = 'signal_normalized'\n",
    "    elif normed==False:\n",
    "        y = 'signal'\n",
    "    for df,color in zip([spectra_abs, spectra_fluo], ['bone_r', 'mako_r']):\n",
    "            df = df.reset_index()\n",
    "            sns.lineplot(data=df[(df['treatment']==treatment) & (df['gene']==gene)\n",
    "                                 & (df['hours']<=57) & (df['Wavelength']<=650)],\n",
    "                        x='Wavelength', y=y, hue='t', ci=None, palette=color, legend=False)\n",
    "    plt.ylim(-0.1,1.1)\n",
    "    label_plot_axis()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_peak_timelapse(y, normed=True):\n",
    "    if normed==True:\n",
    "        col = 'signal_normalized'\n",
    "    elif normed==False:\n",
    "        col = 'signal'\n",
    "    \n",
    "    if y=='Absorbance':\n",
    "        df = spectra_abs\n",
    "        plt.legend(handles = [Line2D([0],[0],c='k', lw=2, label='9M urea'),\n",
    "                         Line2D([0],[0],c='k', ls='--', lw=2, label='1X PBS')],\n",
    "               loc='lower left', frameon=False,)\n",
    "\n",
    "    elif y=='Fluorescence':\n",
    "        df = spectra_fluo\n",
    "\n",
    "        \n",
    "    sns.lineplot(data=df[(df['peak']==True) & (df['gene']!='blank') & (df['hours']<=57)], \n",
    "             x='hours', y=col, hue='gene', ci='sd', style='treatment', legend=False,\n",
    "             palette=[colors['amac'],colors['amacV14L'],colors['cgre'],colors['pplu'],colors['av']])\n",
    "    label_plot_axis(y=y, x='Time (hours)', )\n",
    "    plt.xlim(-2.5, 60)\n",
    "    plt.xticks(range(0,60,10), range(0,60,10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_refolding(time='short'):\n",
    "    assert time in ['short','long']\n",
    "    \n",
    "    if time=='short':\n",
    "        rf = pd.read_csv(os.path.join(structure_folder, 'stability_measurements', \n",
    "                                   'refolding_post_urea__20min.csv'), index_col=0)\n",
    "        for gene in ['amac','amacV14L','cgre','pplu','av']:\n",
    "            xdata = rf[rf.gene==gene]['time']\n",
    "            ydata = rf[rf.gene==gene]['fluorescence_normed']\n",
    "            plt.xticks(range(0, 1201, 120), range(0,21,2))\n",
    "            label_plot_axis(y = 'Fluorescence', x = 'Time (min)')\n",
    "            ax = plt.scatter(xdata,ydata, color=colors[gene], label=gene, s=1, alpha=0.5)\n",
    "#             ax.get_legend().remove()\n",
    "        \n",
    "    elif time=='long':\n",
    "        rf = pd.read_csv(os.path.join(structure_folder, 'stability_measurements', \n",
    "                                   'refolding_post_urea__overnight.csv'), index_col=0)\n",
    "        ax = sns.lineplot(data = rf, x = 'time', y = 'fluorescence_normed', hue='gene', ci = 'sd',\n",
    "            palette = [colors[gene] for gene in ['cgre','av','amacV14L','amac','pplu']])\n",
    "        label_plot_axis(y = 'Fluorescence', x = 'Time (hours)')\n",
    "        ax.get_legend().remove()\n",
    "        plt.xticks(range(0,51000,3600), range(0,15))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib.lines import Line2D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_thermosensitivity():\n",
    "    t = pd.read_csv(os.path.join(structure_folder, 'stability_measurements',\n",
    "                                           'temperature__qPCR_raw_data.txt'),sep='\\t', index_col=0)\n",
    "    t = t[t['SampleName'].str.contains('av|pplu|cgre|amac|amacV14L|blank')].copy()\n",
    "    t['Time'] = t['Time']/1000\n",
    "    t['temperature'] = round(t['Temp'])\n",
    "    \n",
    "    melts = pd.read_csv(os.path.join(structure_folder, 'stability_measurements',\n",
    "                                           'temperature__qPCR_melting_temperatures.txt'),sep='\\t', index_col=0)\n",
    "    melts_tm = {'av':melts[(melts['Name']=='av')]['Tm1'].mean(),\n",
    "             'amac':melts[(melts['Name']=='amac')]['Tm1'].mean(),\n",
    "             'amacV14L':melts[(melts['Name']=='amacV14L')]['Tm1'].mean(),\n",
    "             'cgre':melts[(melts['Name']=='cgre')]['Tm1'].mean(),\n",
    "             'pplu':melts[(melts['Name']=='pplu')]['Tm1'].mean(),}\n",
    "    \n",
    "    norm = {'av':t[(t['SampleName']=='av')&(t['temperature']==50)]['465-510'].mean(),\n",
    "            'amac':t[(t['SampleName']=='amac')&(t['temperature']==50)]['465-510'].mean(),\n",
    "        'amacV14L':t[(t['SampleName']=='amacV14L')&(t['temperature']==50)]['465-510'].mean(),\n",
    "                'cgre':t[(t['SampleName']=='cgre')&(t['temperature']==50)]['465-510'].mean(),\n",
    "                'pplu':t[(t['SampleName']=='pplu')&(t['temperature']==50)]['465-510'].mean(),\n",
    "               'blank':t[(t['SampleName']=='blank')&(t['temperature']==50)]['465-510'].mean()}\n",
    "    \n",
    "    t['normed_fluo'] = t[['465-510', 'SampleName']].apply(lambda x: x[0] / norm[x[1]], axis=1)\n",
    "    \n",
    "    sns.lineplot(data=t[(t['SampleName']!='blank') & (t['Temp']>50)], x='Temp', y='normed_fluo',#y='465-510', \n",
    "                 hue='SampleName', ci='sd', \n",
    "                 palette=[colors['amac'],colors['amacV14L'],colors['cgre'],colors['pplu'],colors['av']])\n",
    "    for gene in ['amac', 'av', 'cgre', 'pplu', 'amacV14L']:\n",
    "        plt.axvline(melts_tm[gene], color=colors[gene], ls=':')\n",
    "    label_plot_axis(y='Fluorescence', x='Temperature (Celsius)', )\n",
    "    \n",
    "    plt.legend(handles = [Line2D([0],[0],c=colors['amacV14L'], lw=2, label='amacGFP:V14L'),\n",
    "                         Line2D([0],[0],c=colors['amac'], lw=2, label='amacGFP'),\n",
    "                         Line2D([0],[0],c=colors['cgre'], lw=2, label='cgreGFP'),\n",
    "                          Line2D([0],[0],c=colors['pplu'], lw=2, label='ppluGFP'),\n",
    "                         Line2D([0],[0],c=colors['av'], lw=2, label='avGFP')],\n",
    "               loc='lower left', frameon=False,)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# amacGFP v amacGFP:V14L"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_mutposeffect(data, y_column, x_val, x_column='position', s_column='mutation'):\n",
    "    df = data[data[x_column]==x_val]\n",
    "    return list(df[y_column]), list(df[s_column])\n",
    "\n",
    "def set_x_coordinates(list_of_yvals, list_of_labels, y_distance_threshold, center_x, x_distance):\n",
    "    '''Tries to find non-overlapping x coordinates for text-swarmplot'''\n",
    "    yvals_to_labels = dict(zip(list_of_yvals, list_of_labels))\n",
    "    list_of_yvals = sorted([x for x in list_of_yvals if str(x)!='nan'])\n",
    "    list_of_xvals = [center_x for x in list_of_yvals]\n",
    "    \n",
    "    ref_y = 0\n",
    "    multiplier = 1\n",
    "#     original_x_distance = x_distance\n",
    "    current_x_spacer = x_distance\n",
    "    \n",
    "#     print(list_of_yvals)\n",
    "    \n",
    "    for i in range(1, len(list_of_yvals)):\n",
    "\n",
    "        if abs(list_of_yvals[i] - list_of_yvals[ref_y]) < y_distance_threshold:\n",
    "#             print(i, 'too close')\n",
    "            list_of_xvals[i] += multiplier*current_x_spacer\n",
    "            multiplier *= (-1)\n",
    "            if multiplier== 1:\n",
    "#                 print(i, 'multiply -1')\n",
    "                current_x_spacer+=x_distance\n",
    "        else:\n",
    "            ref_y = i\n",
    "            current_x_spacer = x_distance\n",
    "#             print(i, 'new ref')\n",
    "\n",
    "    return list_of_xvals, list_of_yvals, [yvals_to_labels[y] for y in list_of_yvals]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Extant vs non-extant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_extant_vs_nonextant_muts(gene, dataset=data_aa, y_axis='brightness', palette='mako', n=5):\n",
    "    c = ListedColormap(sns.color_palette(palette, 256))\n",
    "    df = dataset[dataset.gene==gene+'GFP']\n",
    "    df['f_extant'] = df[['n_mut', 'n_mut_extant']].apply(lambda x:\n",
    "                                            'all' if x[0]==x[1] else 'none' if x[1]==0 else 'some', axis=1)\n",
    "    \n",
    "    plot_half_violin([df[(df['n_mut']==i) & (df['f_extant']=='none')][y_axis] for i in range(1,n)], \n",
    "                     side='left', \n",
    "                     color=[c(80),c(80)], alpha=1, widths=0.8, chonkylines=True)\n",
    "\n",
    "    plot_half_violin([df[(df['n_mut']==i) & (df['f_extant']=='all')][y_axis] for i in range(1,n)], \n",
    "                     side='right', alpha=1, \n",
    "                     color=[c(180), c(180)], widths=0.9, chonkylines=True)\n",
    "#     plt.ylim(df[y_axis].min() - 0.25, df[y_axis].max() + 0.1)\n",
    "\n",
    "    label_plot_axis(x = 'Number of mutations', t= gene+'GFP')\n",
    "    plt.xticks(range(n-1), range(1,1+n))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compare_against_extant(gene, dataset=data_aa, nok=-0.2, overlap_n=10, **kwargs):\n",
    "#     muts_all = set(dataset[singles_mask & eval(gene+'_mask')]['quasipos_genotype'])\n",
    "#     muts_nok = set(dataset[singles_mask & eval(gene+'_mask') \n",
    "#                        & (dataset.measured_effect < nok)]['quasipos_genotype'])\n",
    "#     namekey = {'amac':'GFPxm'}\n",
    "    gene_name = namekey[gene]\n",
    "    gene_wt = get_wt_states(gene_name)\n",
    "\n",
    "    muts_all = {x for x in eval('siffects_'+gene+'_log')}\n",
    "    muts_all = {x[0] + str(pseudopos_to_nativepos[int(x[1:-1])][genekey[gene]]) + x[-1] for x in muts_all}\n",
    "    muts_all = muts_all | {x[0:-1] + x[0] for x in muts_all}\n",
    "    muts_all = {str(nativepos_to_quasipos[int(x[1:-1])][gene_name]) + x[-1] for x in muts_all}\n",
    "    \n",
    "    muts_nok = {x for x in eval('siffects_'+gene+'_log') if eval('siffects_'+gene+'_log')[x] < nok}\n",
    "    muts_nok = {str(pseudopos_to_nativepos[int(x[1:-1])][genekey[gene]]) + x[-1] for x in muts_nok}\n",
    "    muts_nok = {str(nativepos_to_quasipos[int(x[:-1])][gene_name]) + x[-1] for x in muts_nok}\n",
    "\n",
    "    for extant in wts:\n",
    "        if extant=='RrGFP' or extant==namekey[gene]:\n",
    "#             print(extant)\n",
    "            pass\n",
    "        \n",
    "        else:\n",
    "            extant_ok = get_wt_states(extant)\n",
    "            extant_ok = {x for x in extant_ok if x not in gene_wt}\n",
    "                \n",
    "#             print(extant, m)\n",
    "            overlap = {x for x in extant_ok if x in muts_all}\n",
    "            if len(overlap) > overlap_n:\n",
    "                nok_in_gene = {x for x in overlap if x in muts_nok}\n",
    "                aa_id = 100 - pairwise[frozenset([extant, namekey[gene]])]\n",
    "\n",
    "                plt.scatter(aa_id, len(nok_in_gene) / len(overlap), color=colors[gene],**kwargs)\n",
    "#             print(gene,extant, len(overlap))        \n",
    "#     return x, y"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
